### DISCUSSION TITLE
Feedback

### DISCUSSION THREAD

**Discussion by ValueRaider** [ValueRaider/yfinance-cache#39] [Q&A]
> I am not too satisfied with the way yahoo_cache is operating https://github.com/ranaroussi/yfinance/discussions/1665#discussioncomment-6745265

@rickturner2001 would you mind elaborating? I'm receptive to criticisms and suggestions.

**Comment by rickturner2001** [comment]
I believe you've done an exceptional job. What I intended to convey is that the integration of this might not align seamlessly with an existing yfinance project. To illustrate, if I were to incorporate caching into a project that utilizes yfinance, my anticipation would be to execute something along these lines:

```python
import some_package from some_lib
import yfinance as yf

# Cached session instance
config = {...}
session = CachedSession(config=config)

# code ...
yf.download(..., session=)
```

In essence, the only task I would need to undertake here would be generating the session instance, after which the program would maintain its usual functionality.

In summary, my concern pertains to the manner in which the API harmonizes with yfinance, rather than any issues with the implementations themselves.

**Reply by ValueRaider** [reply]
I don't understand. Are you discussing the scenario of yfinance code that already uses a customised requests_cache session, where user is considering switch to yfinance_cache? yfinance_cache doesn't require requests_cache.

Or is it that `download` isn't implemented?

**Reply by rickturner2001** [reply]
You know, as I think about it further, I'm beginning to wonder if the absence of threading is what's bugging me, particularly when it comes to grabbing data for multiple tickers simultaneously. But then again, I could be off the mark in understanding how exactly yfinance_cache is functioning.

To reiterate, I don't have any specific suggestions to make the API better. My concern stems from its reliance on yfinance as a foundation, rather than seamlessly coexisting  with it. For instance, it would be great if we could establish a cache session and seamlessly link it with a yfinance batch download.

**Reply by ValueRaider** [reply]
Ah. Caching isn't certified thread-safe, because need to cache exchange schedules for performance reasons, so potential problem when downloading multiple tickers on same exchange. So not implementing `download`. It could be made threadsafe but I don't see benefit - the data is already cached most-of-time, sequential fetches already fast enough to hit Yahoo rate limits, and Python is single-threaded (because GLI).

This caching works different to requests_cache. Doesn't cache URLs, only the data. But it understands the data, knows when and what to fetch.

Maybe README needs tweaking.

**Reply by rickturner2001** [reply]
I understand the concept, but I'm a bit puzzled as to why caching the URL isn't a feasible approach here. How can you determine whether something being retrieved has been previously fetched, enabling you to potentially provide the cached outcome instead of initiating a redundant fetch?

Regardin Python's GIL limitations, sure, we cannot really achieve true parallelism, but since we're talking about  I/O-bound operations, Multi-threading in Python still allows for concurrency. Woulnd't you see that as a plausible option?

**Reply by ValueRaider** [reply]
> but I'm a bit puzzled as to why caching the URL isn't a feasible approach here

How to handle stock splits, and dividends if you use adjusted prices? Is feasible if you're OK wiping the cache and re-fetching everything. Then how do you update current active trading interval?

> How can you determine whether something being retrieved has been previously fetched

... it's in the cache. A better question would be: when does (some of) cache data need to be refetched?

> talking about I/O-bound operations, Multi-threading in Python still allows for concurrency. Woulnd't you see that as a plausible option?

It's a cost-benefit tradeoff. Benefit = what you describe, maybe, in some scenarios. Cost = development time.

**Reply by ValueRaider** [reply]
Does this improved README explain sufficiently how cache works? https://github.com/ValueRaider/yfinance-cache/pull/38

**Reply by rickturner2001** [reply]
I see. Yes, the new README clearly points out the cababilities and limitations of the API

**Reply by ValueRaider** [reply]
Given your interest in `download`, I've implemented it in #41. Actually wasn't difficult, but not stress-tested it yet.
